core/src/outstanding_requests.rs

```diff
@@ -0,0 +1,166 @@
+use crate::request_response::RequestResponse;
+use lru::LruCache;
+use rand::{thread_rng, Rng};
+use solana_ledger::shred::Nonce;
+use solana_sdk::timing::timestamp;
+
+pub const DEFAULT_REQUEST_EXPIRATION_MS: u64 = 60_000;
+
+pub struct OutstandingRequests<T, S>
+where
+    T: RequestResponse<Response = S>,
+{
+    requests: LruCache<Nonce, RequestStatus<T>>,
+}
+
+impl<T, S> OutstandingRequests<T, S>
+where
+    T: RequestResponse<Response = S>,
+{
+    // Returns boolean indicating whether sufficient time has passed for a request with
+    // the given timestamp to be made
+    pub fn add_request(&mut self, request: T) -> Nonce {
+        let num_expected_responses = request.num_expected_responses();
+        let nonce = thread_rng().gen_range(0, Nonce::MAX);
+        self.requests.put(
+            nonce,
+            RequestStatus {
+                expire_timestamp: timestamp() + DEFAULT_REQUEST_EXPIRATION_MS,
```

Comment:
maybe like the other function this one also better take `now: u64` as an argument rather than calling `timestamp()` itself.